### ISSUE TITLE
Test-generator: operations from blocks

### ISSUE THREAD

**Issue by protolambda** [ethereum/eth2.0-specs#927]
Since the generators are now (waiting for #851 merge) integrated into the PySpec and hosted in this repository, we can push for much more completeness in YAML testing.

This issue is to track the progress of providing test-cases for block-operations ("transactions" previously).
Suggestions for test-cases are welcome. And better, open up a PR with implementation (once #851 is ready)

- [x] Deposits
  - [x] Valid new deposit
  - [x] Valid top-up
  - [x] Invalid: deposit index
  - [x] Invalid: deposit proof
  - [x] Valid new deposit but invalid signature (proof of possession)
- [x] Proposer slashings
  - [x] Valid proposer slashing
  - [x] Invalid proposer index
  - [x] Invalid: different epoch
  - [x] Invalid: Same block headers
  - [x] Invalid: Non-slashable proposer (not activated) [TODO: correct, maybe just slash them anyway, 
since they signed it?]
  - [x] Invalid: Non-slashable proposer (slashed)
  - [x] Invalid: Non-slashable proposer (withdrawn)
  - [x] Invalid: Invalid signatures (either header, or both)
- [x] Attester slashings
  - [x] Invalid: Same attestation
  - [x] Valid: double vote
  - [x] Valid: surround vote
  - [x] Invalid: no double or surround vote
  - [x] Invalid: either attestation has invalid indexed form (custody bit indices wrong)
  - [x] Invalid: no slashable indices
  - [x] Invalid: invalid signatures (either attestation, or both)
- [x] Attestations (inclusion)
  - [x] Valid attestation
  - [x] Valid previous epoch
  - [x] Invalid: shard out of range
  - [x] Invalid: incorrect source epoch (previous and current epoch)
  - [x] Invalid: incorrect source root (previous and current epoch)
  - [x] Invalid: non-zero crosslink root
  - [x] Invalid: incorrect previous crosslink root with state (previous)
  - [x] Invalid: source root is target root
  - [x] Invalid: attestation references non-zero custody bitfield
  - [x] Invalid: current and previous root switched
  - [x] Invalid: mismatch in participants/custody bits
  - [x] Invalid: invalid signature
- [x] Voluntary Exits
  - [x] Valid exit
  - [x] Exit queue
  - [x] Invalid: out of range validator index
  - [x] Invalid: not active (activation phase)
  - [x] Invalid: already exited
  - [x] Invalid: voluntary exit is in the future
  - [x] Invalid: exit is too soon
  - [x] Invalid: invalid signature
- [x] Transfers
  - [x] Valid transfer from not yet activated validator
  - [x] Valid transfer from withdrawable validator
  - [x] Valid transfer, above effective balance
  - [x] Valid transfer, above effective balance, just fee
  - [x] Invalid: active, transfer past effective balance
  - [x] Invalid: transfer past effective balance, by fee
  - [x] Invalid: balance less than transfer amount
  - [x] Invalid: balance less than fee amount
  - [x] Invalid: transfer slot is not the current state slot
  - [x] Invalid: withdrawal credentials are incorrect
  - [x] Invalid: sender account is dust afterwards
  - [x] Invalid: recipient account is dust afterwards
  - [x] Invalid: incorrect signature

staged tests (i.e. need review/update):
- [ ] Invalid: incorrect previous crosslink root with state (current epoch)

**Comment by JustinDrake** [comment]
Can this be closed Proto?

**Comment by protolambda** [comment]
Yes. I'm running coverage reports on ZRNT, and it's looking good for most parts of the spec, but will open another issue with remaining tests to be written later this week.
